### DISCUSSION TITLE
How to refetch query with params

### DISCUSSION THREAD

**Discussion by skryvytskyi** [TanStack/query#351] [General] [answered]
For example I created a query 
`  const {
    data = {
      result: [],
    },
  } = useQuery('data_sources', getDataSources);`

This data used for select component and I wonna refetch it with `{ search: value }` param.
Previously we had a `variables` params inside `refetch`, now it's removed.

**Comment by arb000r** [comment]
I was wondering this as well

**Comment by tannerlinsley** [comment, marked as answer]
Recommended: Turn your variables into a permanent query key part and just update them with state or props

```js
useQuery(['data_sources', { variables }], getDataSources)
// or
useQuery(['data_sources', variables], getDataSources)
```

**Reply by BertCatsburg** [reply]
Ok, so you have a form and type some values. And those values should be used in the refetch. 
If I understand correctly you recommend to do this as follows:
1) Pick up your values in some onSubmitHandler (like react-hook-form)
2) Put them from the onSubmitHandler to some state (useState)
3) useQuery reads them from State and executes the query. 

This is 3 steps to do a simple retrieve. 

Why not execute the refetch from onSubmitHandler like:
```js
refetch({email: "myemail@email.com", password: "12345"});
```
This would save the useState handling.

**Reply by TkDodo** [reply]
because useQuery is declarative, and `refetch` is a way to do an imperative fetching. refetch bypasses staleTime / enabled prop and just re-fetches the current query with the current fetch-fn.

refetching with different parameters would need to create a new query key, but which one? we probably don't want to override the current query key with data that comes from fetching with different parameters. what if your queryFn closures over props, how would that work?

imo, "refetching with parameters" like that just doesn't fit very well with the mental model of react-query.

**Reply by BertCatsburg** [reply]
Hello TkDodo,

Thanks for your explanation and technically you are absolutely right. 
But it is fairly standard behaviour that you have a form, fill in some fields and want to do a query with those fields as input.
For example a login-form where you input email,password and want to get the JWT-token back. 

Currently this requires a submitHandler, a useState, a useEffect and a useQuery. That's a lot of code for doing a simple query. 
If somebody can write the above functionality in a better and more simple way (and put it in the docs!!) it would clarify a lot.

**Reply by TkDodo** [reply]
login is indeed s mutation, but I get that with forms, there is a bit of overhead involved: you need local state, and when the user clicks apply, you have to lift the state up and put that in the query key. I use the url for that, and plan to blog about it soon. But you certainly don’t need a useEffect - what would that be for please?

**Reply by BertCatsburg** [reply]
Correct, I managed it without the useEffect. 
And i took a while to understand why this should be a useMutation instead of a useQuery while in fact it is only doing a query (the accessToken). 
React-Query is a fantastic library and it deserves better docs and more real-life examples.

**Reply by TkDodo** [reply]
It doesn’t get/query the access token though - it _creates_ the access token. Big difference :)

**Reply by BertCatsburg** [reply]
I understand. But the creation is done on the server/API, not something the client sees. 
What if you do a useQuery and the server/API does some logging/tracing (also a create) record to a logging database. 
Then also a create is done. 
What's the difference?

**Reply by TkDodo** [reply]
the difference is that queries can happen out of nowhere and should be idempotent and have no side-effects on the server. the query can happen when a window is focused or when the browser re-connects to the network. Are you okay with your request happening often, then a query is fine. Think about an email that is sent, like a one-time token mail, every time this "request" is made. That is likely not a query, but a mutation. mutation because it creates a side-effect _on the server_.

if you think that creating a token is not a side-effect, and you can do that a lot and also discard results (react-query will discard results when the query is made twice in succession, and queries are supposed to be idempotent), then you can also use a query.

**Reply by BertCatsburg** [reply]
Aha, now I understand. Thanks.

**Reply by cmacdonnacha** [reply]
@TkDodo I have a similar question but a different use case if you wouldn't mind clarifying what should be used:

1. User is presented with a table of cars: GET `/cars`
2. User selects a car manufacturer that they want to filter by: GET `/cars?make=ford`

I currently have this query which fetches all cars:

`const { status, data } = useQuery('carsData', () => fetch('/cars');`

Would something like this work where I use the filter as part of the key? 
`  const [filter, setFilter] = useState('');`
`  const { status, data } = useQuery(['carsData', filter], () => fetch('/cars?make=${filter}');`

Or is there a better way? Perhaps have a seperate query altogether? 

I'm guessing my use case would not be seen as a mutation right? Compared to the login example given above?

**Reply by TkDodo** [reply]
your case is definitely a query, because you are just fetching different kind of data. It is also recommended to add all dependencies of the queryFn, in your case the filter, to the queryKey, so this is definitely a good pattern. It means that you'll get an automatic refetch and a new cache entry when you call `setFilter`. Other advantages include data cached by filter criteria, so once you go back to filtering all cars, you still have the data in the cache, and you'll instantly get the data and a background refetch. I've written about this pattern in my blog:

https://tkdodo.eu/blog/practical-react-query

**Reply by cmacdonnacha** [reply]
Excellent thanks @TkDodo - a very relevant article for my use case indeed!

**Reply by cmacdonnacha** [reply]
One last one, what would the query look like if I had both filtering and pagination? How can you pass more than one dependency? Would it be something as simple as this?

`const { status, data } = useQuery(['carsData', filter, page], () => fetch('/cars?page=${page}&make=${filter}');`

**Reply by TkDodo** [reply]
that would work fine, you can pass as many dependencies as you want, filter, paging, sorting, limit, ... you can also create an object if you like that better, but don't have to (react-query will make sure that objects in keys are stable):

`useQuery(['carsData', { filter, page }]`

**Reply by cmacdonnacha** [reply]
Excellent thanks @TkDodo

**Reply by cmacdonnacha** [reply]
Hey @TkDodo , apologies to be adding to this again but I was wondering how you can use `react-query` to handle errors for a similar scenario:

1. User has 3 buttons which when clicked will re-fetch some filtered data. e.g Filter buttons options are "Ford", "Audi" and "Tesla"
2. "Ford" filter button is selected by default an has "active" styling to show this.
3. User clicks "Tesla" filter button and `react-query` makes a request out to retrieve only tesla cars:
`setFilter('tesla');`
`const { status, data } = useQuery(['carsData', filter], () => fetch('/cars?make=${filter});`
4. Button styling changes to show "Tesla" is now the selected filter.
5. This time the user clicks back to "Ford" but an error has occurred with the request.

How can I ensure in this case that the "Ford" button is not seen as the "Selected" button? I know in `useMutation` there an `onError` callback where you can do a rollback. Is there something equivalent that can be done for `useQuery`?

**Reply by BertCatsburg** [reply]
Hello @cmacdonnacha, if you could put a complete but small test-application with your question in a Repo on Github, others could clone it and try to find a solution for you. I'm happy to try to help you out here, as long as you make it easy for me. :-)

**Reply by TkDodo** [reply]
useQuery also has an onError callback, but I don't quite get why you would want to revert here. The way I'm seeing it, the user has selected `Ford` and an error has occurred. Now you can choose to display the stale Ford data from the cache that you already have, or you can display an Error like "Could not fetch Ford cars" instead (or you can even show both!). But in no way would I expect the filter criteria that the user actively chose to go back to Tesla ... or am I getting the scenario wrong?

**Reply by cmacdonnacha** [reply]
Actually yes, you're right - if an error occurs when the user switches from "Ford" to "Tesla", it would actually be a better user experience to show "Tesla" as the selected filter button with an error to tell the user that something went wrong while choosing this particular filter. Not sure why I had "revert" in my head when it's not needed here at all and those "optimistic update" scenarios should only be done when updating or creating something while using `useMutation`, not so much with `useQuery`. Thanks again for your patience and constant help in the `react-query` community, it doesn't go unnoticed!

**Reply by mogaozq** [reply]
what if I use `useMutation` in the case caching data is not needed?( using mutation to get data)

**Reply by TkDodo** [reply]
mutations don’t share data between components, have no cache and cannot do background updates. I see no reason to do this :)

**Reply by taisuke-j** [reply]
I just encountered a similar situation and follow the tips above. Just one more thing to add here - I also used `enabled` option to make sure the refetch gets triggered only when an id is selected/changed by user's click

```
const [ownderId, setOwnderId] = useState<string>("");

...

const onOwnderClick = useCallback((id: string) => {
  setOwnderId(id);
}, []);

...

return useQuery(["cars", ownderId], fetchCars, {
  enabled: !!ownderId, // THIS
   ...
});
```

**Reply by An-Huynh-Certivity** [reply]
@tannerlinsley 
I am currently using a setup similar to this
```
const Parent = () => {
    const [currentInput, setCurrentInput] = useState(1)
    {isLoading, error, data} = useFetch(['key', currentInput], fetchData(currentInput))
    
    return (
        <div>
            <Child data={data}/>
            <button onClick={() => setCurrentInput(1)}>input1</button>
            <button onClick={() => setCurrentInput(2)}>input2</button>
        </div>
    )
}
```
And everything works fine for the first refetch:  
- setCurrentInput() triggers a refetch of the document
- isLoading === true unloads the Child component and once the data has been fetched, displays the new data

Unfortunately the second refetch will not change isLoading, since the result will be cached already. Therefore the Child component will not unload and since data is not a state variable, Child does not know that data has changed and will still display the old data.

I am currently using a somewhat ugly workaround by forcing rerender like this:
```
const Parent = () => {
  const [currentInput, setCurrentInput] = useState(1);
  const [isRefreshing, setIsRefreshing] = useState(false);
  const { isLoading, error, data } = useFetch(
    ["key", currentInput],
    fetchData(currentInput)
  );

  function refreshChild() {
    setIsRefreshing(true);
    setTimeout(() => setIsRefreshing(false), 5);
  }

  return (
    <div>
      {!isLoading && !error && !isRefreshing && data && <Child data={data} />}
      <button
        onClick={() => {
          setCurrentInput(1);
          refreshChild();
        }}
      >
        input1
      </button>
      <button
        onClick={() => {
          setCurrentInput(2);
          refreshChild();
        }}
      >
        input2
      </button>
    </div>
  );
};
```

However this approach does not seem super elegant and I was wondering whether there are patterns that better deal with this problem.

**Reply by TkDodo** [reply]
> Therefore the Child component will not unload and since data is not a state variable, Child does not know that data has changed and will still display the old data.

I'm not sure I understand the problem, or what result you would want to achieve. `Child` will always get the latest `data`. I can only assume that the `Child` component maybe takes the `data` prop and puts it into local state or so? Judging from the fact that you would like to achieve an "unmount" ?

If that's the case, for one, I'm thinking that's a bit of an anti pattern, but you can also force a component to re-mount by passing a `key` prop. So:

```
<Child data={data} key={currentInput} />
```

will likely work. I've blogged about this here: https://tkdodo.eu/blog/putting-props-to-use-state

If I'm wrong with the assumption, please elaborate on your issue.

**Reply by An-Huynh-Certivity** [reply]
You are absolutely correct in that the `Child` component loads the props into local state! I have taken a look at your blog article which I found very helpful and has solved my issue :)

**Reply by CharanRJS** [reply]
@TkDodo I have a problem here sir , I hope you would help me. I just started learning USEQUERY lib. still a beginner, writing a code that calls a backend api, that api has ```itemsperpage``` and ```page``` params, and that api needs to call everytime I change a number. I have refetch but its calling again with same default params. Please help me

**Reply by TkDodo** [reply]
I think you want to use an infinite query here.

**Reply by CharanRJS** [reply]
@TkDodo Hi Sir , Thanks for speedy reply , sir Actually Im using react table on other hand , can you suggest me any tutorial you know.Im trying to use ```useInfinityQueries``` but its not working with ```react-table```

**Reply by beingsandthings** [reply]
@TkDodo 
Hello, can i ask a question? 
I want to create a download button where the user can click on a file and get a "fileId" value and download it with the value. However, I'm having some issues with useQuery to download the file. It downloads whenever the state it depends changes without click event.

To solve this problem, I've tried using refetch, useState, and useEffect to set a "fileId" value and trigger a download when the user clicks the button. It works well, but I feel like there might be a better way to do it.

I'm trying putting the "fileId" value in the queryKey as u said. 

1. But I'm wondering if I should use useMutation instead of useQuery, since downloading a file could be seen as generating a file to be handed over, that is, having side-effects on the server.

2. Or should i use url to solve this problem? You'v mentioned using URLs in the situation where I need local state and have to lift the state up and put it in the query key.

**Reply by TkDodo** [reply]
I would `useMutation` here, especially if the file is generated every time you click the button.

**Reply by beingsandthings** [reply]
> I would `useMutation` here, especially if the file is generated every time you click the button.


Thank u for answering! I have one more question. What if the file is not generated but retrieved from db..? Do u think `useMutation` is still available for this case?

**Comment by tannerlinsley** [comment]
Login Is not a query, it’s a mutation. Use useMutation instead.

Tanner Linsley

**Comment by mrdulin** [comment]
@TkDodo Hi, can you take a look at my issue? Thanks. I hit this issue by using `react-query` on day 1. I've read most of the documentation and all of the examples but still doesn't find a solution.

My page has a search form and table with a paginator. So I have two states: `searchCondition` and `pagination`. When the user clicks the paginator, the pagination state will be changed so that a new `channelList` query will be made. That's ok. The search form includes two input form items and a "Query" button. The `channelList` query also can be triggered by the user click event.

Currently:

* Send the query on the component mount. - done!
* Made a new query when pagination state changed - done!
* User types something in the input form item(this will cause the searchCondition state changed) and clicks the "Query" button, then made a new query - done!

**The problem is:** When user types something in inputs and click the "Query" button, the `searchCondition` state changed and the `handleSearchFormSubmit` is triggered. I want to reset pagination to the initial value, then call `channelListQuery.refetch()`. But it sends two queries: one by `channelListQuery.refetch()` and another by `useEffect(() => {}, [pagination])`, that's not expected.

For example, I am currently on page 3, when the user types something and clicks the "Query" button.  The `handleSearchFormSubmit ` event handler was executed. There are two queries:

```
?current=3&channelName=xxx // triggered by channelListQuery.refetch()
?current=1&channelName=xxx // triggered by useEffect
```

It's correct. Because the `resetPagination()` will not update `pagination` immediately, and `pagination` used by `refetch()`  is the previous value.

```js
const { searchCondition, setSearchCondition } = useSearchCondition(initialSearchCondition);
  const { pagination, setPagination, resetPagination } = usePagination();
  const channelListQuery = useQuery(
    ['channelList', pagination, searchCondition],
    async () => {
      const getChannelListParams = omitFalsyFields({
        currentPage: pagination.current,
        pageSize: pagination.pageSize,
        ...searchCondition,
      });
      const { code, result } = await apis.channel.getChannelList(getChannelListParams);
      if (code === '0') {
        return result;
      }
    },
    { initialData: { totalItem: 0, resultList: [] }, refetchOnWindowFocus: false, keepPreviousData: true, enabled: false, cacheTime: 0 },
  );

  useEffect(() => {
    channelListQuery.refetch();
  }, [pagination]);

  const handleSearchFormSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    resetPagination();
    channelListQuery.refetch();
  };

// ...
pagination={{
            total: channelListQuery.data.totalItem,
            ...pagination,
            showQuickJumper: true,
            showSizeChanger: true,
            onChange: (page: number, pageSize?: number) => {
              const nPagination = { current: page, pageSize: pageSize || paginationInitialState.pageSize };
              setPagination(nPagination);
            },
          }}
```

Maybe `refetch()` with parameters inside `onChange` event of paginator and `handleSearchFormSubmit` event handler instead of listening to the changes of the state could solve this situation? That's why I came to this issue.  I solved it this way before I used `react-query`. Like this:

```js
  useEffect(() => {
    channelListQuery.refetch();
  }, []);

  const handleSearchFormSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    channelListQuery.refetch({...paginationInitialState, ...searchCondition});
  };
//  pagination
onChange: (page: number, pageSize?: number) => {
              const nPagination = { current: page, pageSize: pageSize || paginationInitialState.pageSize };
              setPagination(nPagination);
              channelListQuery.refetch({...nPagination, ...searchCondition});
            },

```

So how does `react-query` handle this situation?

**Reply by TkDodo** [reply]
I answered the question on StackOverflow: https://stackoverflow.com/a/71093384/8405310

**Reply by cliffordfajardo** [reply]
> My page has a search form and table with a paginator. So I have two states: searchCondition and pagination. When the user clicks the paginator, the pagination state will be changed so that a new channelList query will be made. That's ok. The search form includes two input form items and a "Query" button. The channelList query also can be triggered by the user click event.


I was struggling with this today as well; I though maybe `refetch` would help, but unfortunately, you can't pass down arguments to it; even if that did work, I would need to pass my `refetch` function down as props

I decided to go with the updating the [URL solution](https://stackoverflow.com/a/71093384/8405310) that @TkDodo mentioned in the comment above, which has the added benefit of
- shareable URLs
- not needing to pass down props 


![CleanShot 2022-05-26 at 11 20 31@2x](https://user-images.githubusercontent.com/6743796/170551612-b7b8ef59-6c8e-4f49-8cd6-0845f88e2662.png)

**Reply by TkDodo** [reply]
this is a perfect use case for it 👍

**Reply by adriannecris** [reply]
But how do you handle debouncing if your enabled option is dynamic?

**Comment by 1mehdifaraji** [comment]
```typescript
  const {
    refetch: orderInvoiceRefetch,
  } = useQuery(
    ["order-invoice"],
    (orderId) => axios
        .get(`/orders/invoice?order_id=${orderId}`)
        .then(({ data })) => data),
    { enabled: false }
  );
```
I need to pass `id` param to the `orderInvoiceRefetch` refetch function and I don't want to use website url `query`.
```typescript
orderInvoiceRefetch({
   queryKey: id,
});
```
How can I pass a parameter to the `refetch` function?

**Reply by vossmalte** [reply]
You don't: https://tkdodo.eu/blog/react-query-fa-qs#how-can-i-pass-parameters-to-refetch

**Reply by 1mehdifaraji** [reply]
Sad to hear, However I did it with react's own `state` with `enabled` option of react query like this:
```typescript
const [orderId, setOrderId] = useState<string | null>(null);

const { data } = useQuery(["order-user", orderId], ({ queryKey }) =>
      axios
        .get(`/orders/user?order_id=${queryKey[1]}`)
        .then(({ data }: { data: OrderUserResponse }) => data),
    { enabled: !!orderId }
);
```

**Reply by TkDodo** [reply]
that's the idiomatic way - why is that sad? Now you don't need to call `refetch` manually - a change in `orderId` will trigger a new fetch automatically. Data is also cached by its inputs, thus avoids race conditions and we can show stale, cached data earlier when switching between orderIds.

**Reply by 1mehdifaraji** [reply]
@TkDodo I couldn't pass arguments to the `useQuery` and had to use react `useState` or the url `query` with`enabled` option to accomplish such a simple thing.

**Reply by vossmalte** [reply]
It is called refetch as it should trigger the same fetch again, so no params.

**Reply by windmaomao273** [reply]
I see. so `orderId` coded inside the key essentially is like `useEffect` parameter?

**Reply by TkDodo** [reply]
it’s similar, yes

**Reply by windmaomao273** [reply]
@TkDodo thank you, i am using this approach now. Quite functioning.
